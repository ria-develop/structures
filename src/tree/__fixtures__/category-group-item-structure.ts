import { generatorToArray } from '../utils';

export const grouping: ['category', 'group', 'item'] = ['category', 'group', 'item'];

export type DataSource = {
  category: string;
  categoryId: string;
  group: string;
  groupId: string;
  item: string;
  itemId: string;
  value: number;
};

function* categoryFactory(numCategories: number, numGroups: number, numItems: number): Generator<DataSource> {
  const total = numCategories * numGroups * numItems;
  let current = 0;
  let categoryId = 1;
  let groupId = 1;
  let itemId = 0;
  while (current < total) {
    current++;
    if (!(current % 4)) {
      categoryId++;
      groupId = 0;
      itemId = 0;
    }
    if (!(current % 2)) {
      groupId++;
      itemId = 0;
    }
    itemId++;

    yield {
      category: 'Category ' + categoryId,
      categoryId: categoryId + '',
      group: 'Group ' + groupId,
      groupId: groupId + '',
      item: 'Item ' + itemId,
      itemId: itemId + '',
      value: 100
    };
  }
}

export const getData = (numCategories = 2, numGroups = 2, numItems = 2): DataSource[] =>
  generatorToArray<DataSource>(categoryFactory(numCategories, numGroups, numItems));
