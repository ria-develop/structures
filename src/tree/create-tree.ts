import { BaseObject, Id, LeafNode, NodeType, ParentNode, RootNode, TreeNode } from './types';

const isLeaf = (type: NodeType): type is NodeType.LEAF => type === NodeType.LEAF;
const isParent = (type: NodeType): type is NodeType.PARENT => type === NodeType.PARENT;
const isRoot = (type: NodeType): type is NodeType.ROOT => type === NodeType.ROOT;

export const isNodeLeaf = <T extends BaseObject>(node?: TreeNode<T>): node is LeafNode<T> =>
  !!node && isLeaf(node.type);

export const isNodeParentLike = <T extends BaseObject>(node?: TreeNode<T>): node is ParentNode<T> | RootNode<T> =>
  !!node && 'children' in node;

export const isNodeParent = <T extends BaseObject>(node?: TreeNode<T>): node is ParentNode<T> =>
  !!node && isParent(node.type);

const createLeafNode = <T extends BaseObject>(valueField?: keyof T, data?: T): LeafNode<T> => ({
  data,
  get value() {
    return data[valueField];
  },
  type: NodeType.LEAF
});

const createParentNode = <T extends BaseObject>(valueField?: keyof T, data?: T): ParentNode<T> =>
  Object.defineProperty(
    {
      children: [],
      data,
      get value() {
        return data[valueField];
      },
      type: NodeType.PARENT
    },
    'data',
    {
      value: data,
      enumerable: false,
      configurable: true
    }
  );

const createRootNode = <T extends BaseObject>(): RootNode<T> => ({
  children: [],
  type: NodeType.ROOT
});

const createNodeByType = <T extends BaseObject>(type: NodeType, valueField?: keyof T, data?: T): TreeNode<T> => {
  if (isLeaf(type)) {
    return createLeafNode(valueField, data);
  } else if (isParent(type)) {
    return createParentNode(valueField, data);
  } else if (isRoot(type)) {
    return createRootNode();
  }
};

const getId = (parentId: Id, key: string | number): Id => {
  let id: Id = parentId[key];
  if (!id) {
    id = {};
    parentId[key] = id;
  }
  return id;
};

function addNodeToParent<T extends BaseObject>(node: TreeNode<T>, parentNode: ParentNode<T> | RootNode<T>): void {
  Object.defineProperty(node, 'parent', {
    value: parentNode,
    enumerable: false,
    configurable: true
  });
  if (isNodeParentLike(parentNode)) {
    parentNode.children.push(node);
  }
}

export type ReduceListToTreeProps<T extends BaseObject> = {
  list: T[];
  root: Id;
  grouping: (keyof T)[];
  nodeById: Map<Id, TreeNode<T>>;
  rootNode: TreeNode<T>;
  createGroupCallback: CreateGroupCallback<T>;
};

function reduceListToTree<T extends BaseObject>({
  grouping,
  list,
  root,
  rootNode,
  nodeById,
  createGroupCallback
}: ReduceListToTreeProps<T>): TreeNode<T> {
  return list.reduce((accumulator: TreeNode<T>, dataItem: T) => {
    let parentId: Id = root;
    let parentNode: TreeNode<T> = accumulator;

    grouping.forEach((group: keyof T, index: number) => {
      const name: T[keyof T] = dataItem[group];
      const id = getId(parentId, name);
      let node = nodeById.get(id);
      if (!node) {
        const nodeType = index === grouping.length - 1 ? NodeType.LEAF : NodeType.PARENT;
        node = createNodeByType(nodeType, group, dataItem);
        if (isNodeParentLike(node) && typeof createGroupCallback === 'function') {
          node = createGroupCallback(node, dataItem, group);
        }

        nodeById.set(id, node);
        if (node && !isNodeLeaf(parentNode)) {
          addNodeToParent(node, parentNode);
        }
      }
      parentId = id;
      parentNode = node;
    });
    return accumulator;
  }, rootNode);
}

export type CreateGroupCallback<T extends BaseObject> = (
  groupNode: ParentNode<T> | RootNode<T>,
  entry: T,
  group: keyof T
) => ParentNode<T> | RootNode<T>;

export function createTree<T extends BaseObject>(
  list: T[],
  grouping: (keyof T)[],
  createGroupCallback?: CreateGroupCallback<T>
): RootNode<T> {
  const root: Id = {};
  //TODO: Create pointer-base id instead of value-base
  const nodeById: Map<Id, TreeNode<T>> = new Map<Id, TreeNode<T>>();
  const rootNode: RootNode<T> = createRootNode();
  return reduceListToTree({ list, root, grouping, nodeById, rootNode, createGroupCallback }) as RootNode<T>;
}
