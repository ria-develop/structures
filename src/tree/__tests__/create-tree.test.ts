import { createTree } from '../create-tree';
import { getData, DataSource, grouping } from '../__fixtures__/category-group-item-structure';

describe('Given category, group, item flat data provided', () => {
  const data = getData();
  let tree;
  it('should return the consistent tree structure', () => {
    tree = createTree<DataSource>(data, grouping);
    expect(tree).toMatchSnapshot();
  });
  it('should give the access to origin data on each level of a tree', () => {
    const [firstOrigin] = data;
    expect(tree).toHaveProperty(['children', 0, 'data'], firstOrigin);
    expect(tree).toHaveProperty(['children', 0, 'children', 0, 'data'], firstOrigin);
    expect(tree).toHaveProperty(['children', 0, 'children', 0, 'children', 0, 'data'], firstOrigin);
  });
});
