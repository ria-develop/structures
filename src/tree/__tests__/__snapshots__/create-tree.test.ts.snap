// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given category, group, item flat data provided should return the consistent tree structure 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "data": Object {
                "category": "Category 1",
                "categoryId": "1",
                "group": "Group 1",
                "groupId": "1",
                "item": "Item 1",
                "itemId": "1",
                "value": 100,
              },
              "type": 2,
              "value": "Item 1",
            },
          ],
          "type": 1,
          "value": "Group 1",
        },
        Object {
          "children": Array [
            Object {
              "data": Object {
                "category": "Category 1",
                "categoryId": "1",
                "group": "Group 2",
                "groupId": "2",
                "item": "Item 1",
                "itemId": "1",
                "value": 100,
              },
              "type": 2,
              "value": "Item 1",
            },
            Object {
              "data": Object {
                "category": "Category 1",
                "categoryId": "1",
                "group": "Group 2",
                "groupId": "2",
                "item": "Item 2",
                "itemId": "2",
                "value": 100,
              },
              "type": 2,
              "value": "Item 2",
            },
          ],
          "type": 1,
          "value": "Group 2",
        },
      ],
      "type": 1,
      "value": "Category 1",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "data": Object {
                "category": "Category 2",
                "categoryId": "2",
                "group": "Group 1",
                "groupId": "1",
                "item": "Item 1",
                "itemId": "1",
                "value": 100,
              },
              "type": 2,
              "value": "Item 1",
            },
            Object {
              "data": Object {
                "category": "Category 2",
                "categoryId": "2",
                "group": "Group 1",
                "groupId": "1",
                "item": "Item 2",
                "itemId": "2",
                "value": 100,
              },
              "type": 2,
              "value": "Item 2",
            },
          ],
          "type": 1,
          "value": "Group 1",
        },
        Object {
          "children": Array [
            Object {
              "data": Object {
                "category": "Category 2",
                "categoryId": "2",
                "group": "Group 2",
                "groupId": "2",
                "item": "Item 1",
                "itemId": "1",
                "value": 100,
              },
              "type": 2,
              "value": "Item 1",
            },
            Object {
              "data": Object {
                "category": "Category 2",
                "categoryId": "2",
                "group": "Group 2",
                "groupId": "2",
                "item": "Item 2",
                "itemId": "2",
                "value": 100,
              },
              "type": 2,
              "value": "Item 2",
            },
          ],
          "type": 1,
          "value": "Group 2",
        },
      ],
      "type": 1,
      "value": "Category 2",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "data": Object {
                "category": "Category 3",
                "categoryId": "3",
                "group": "Group 1",
                "groupId": "1",
                "item": "Item 1",
                "itemId": "1",
                "value": 100,
              },
              "type": 2,
              "value": "Item 1",
            },
          ],
          "type": 1,
          "value": "Group 1",
        },
      ],
      "type": 1,
      "value": "Category 3",
    },
  ],
  "type": 0,
}
`;
